tree.go

```diff
@@ -1266,3 +1266,155 @@ func ToDot(root VerkleNode) string {
 	root.Commit()
 	return fmt.Sprintf("digraph D {\n%s}", root.toDot("", ""))
 }
+
+type SerializedNode struct {
+	Node            VerkleNode
+	CommitmentBytes [32]byte
+	SerializedBytes []byte
+}
+
+// BatchSerialize is an optimized serialization API when multiple VerkleNodes serializations are required, and all are
+// available in memory.
+func (n *InternalNode) BatchSerialize() ([]SerializedNode, error) {
+	// Commit to the node to update all the nodes commitments.
+	n.Commit()
+
+	// Collect all nodes that we need to serialize.
+	nodes := make([]VerkleNode, 0, 1024)
+	nodes = n.collectNonHashedNodes(nodes)
+
+	// We collect all the *Point so we can batch invert the work the projective->affine transformation.
+	pointsToCompress := make([]*Point, 0, len(nodes)*NodeWidth+len(nodes))
+	pointToCompressPerNodeCount := make([]int, len(nodes))
+	for i := range nodes {
+		switch n := nodes[i].(type) {
+		case *InternalNode:
+			pointsToCompress = append(pointsToCompress, n.commitment)
+			count := 0
+			for _, c := range n.children {
+				_, isInternalNode := c.(*InternalNode)
+				_, isLeafNode := c.(*LeafNode)
+				if isInternalNode || isLeafNode {
+					pointsToCompress = append(pointsToCompress, c.Commitment())
+					count++
+				}
+			}
+			pointToCompressPerNodeCount[i] = 1 + count // Internal node commitment + all non-hashed children.
```

Comment:
For an internal node, we add in `pointsToCompress` all the non-hashed children commitments plus its own commitment.

Reply:
won't that make things redundant then? If the child's commitment is included, then it has been modified. If the child is an internal node, then it will also have its children's commitments as well as its own.

Reply:
Good point. I'd have to think a bit about how to reorganize some things here, maybe with an extra `map`.
I'll think about it a bit.

Reply:
OK, [fixed it here](https://github.com/gballet/go-verkle/pull/333/commits/b1e3111371b05643f4ee49ab53830d69889b7878).
Maybe take a look if looks convincing.